File: php-track/php-app/src/Catalog/Product.php
<?php

declare(strict_types=1);

namespace App\Catalog;

use App\Money\Money;
use App\Support\Timestamps;
use DateTimeImmutable;

/**
 * Product is a ShopKit catalog item. It demonstrates composition two ways:
 *   - it USES the Timestamps trait (horizontal reuse, no inheritance), and
 *   - it HOLDS a Money price as a field.
 * Neither is inheritance — Product `extends` nothing.
 */
final class Product
{
    use Timestamps;

    public function __construct(
        public readonly string $sku,
        public readonly string $name,
        private Money $price,
    ) {
    }

    public static function create(string $sku, string $name, Money $price, DateTimeImmutable $now): self
    {
        $product = new self($sku, $name, $price);
        $product->touch($now);

        return $product;
    }

    public function price(): Money
    {
        return $this->price;
    }

    public function reprice(Money $newPrice, DateTimeImmutable $now): void
    {
        $this->price = $newPrice;
        $this->touch($now);
    }
}

File: go-track/go-app/internal/catalog/product.go
// Package catalog holds the Product type from the ShopKit domain.
//
// Product demonstrates composition: it EMBEDS audit.Timestamps (gaining Touch,
// CreatedAt, UpdatedAt) and HOLDS a money.Money price as a field. Neither is
// inheritance — there is no base "Entity" class Product extends.
package catalog

import (
	"time"

	"oop-patterns-learning/go-app/internal/audit"
	"oop-patterns-learning/go-app/internal/money"
)

// Product is a catalog item.
type Product struct {
	audit.Timestamps // embedded: Product.Touch(), Product.CreatedAt, etc. are promoted

	SKU   string
	Name  string
	Price money.Money
}

// NewProduct constructs a Product and stamps its timestamps.
func NewProduct(sku, name string, price money.Money, now time.Time) *Product {
	p := &Product{SKU: sku, Name: name, Price: price}
	p.Touch(now) // promoted method from the embedded Timestamps
	return p
}

// Reprice changes the price and re-stamps UpdatedAt.
func (p *Product) Reprice(newPrice money.Money, now time.Time) {
	p.Price = newPrice
	p.Touch(now)
}
